File: python/leetcode/084.py
class Solution:
    def largestRectangleArea(self, height):
        stack, i, res = [], 0, 0
        h = height + [0]
        while i < len(h):
            if not stack or h[stack[-1]] <= h[i]:
                stack.append(i)
                i += 1
            else:
                t = stack.pop()
                res = max(res, h[t] * (i if not stack else i - stack[-1] - 1))
        return res

File: cpp/leetcode/084.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0);
        int res = 0;
        stack<int> stk;
        for(int i = 0; i < heights.size(); i++){
            if(stk.empty()) stk.push(i);
            while(!stk.empty() && heights[i] < heights[stk.top()]){
                int idx = stk.top(); stk.pop();
                int width = stk.empty() ? i : (i - stk.top() - 1);
                res = max(heights[idx] * width, res);
            }
            stk.push(i);
        }
        return res;
    }
};